### PR TITLE
Lookup resource by name

### PR DESCRIPTION
**Pull request by zachmullen**
For once, I actually managed to keep my commits logically isolated.

* **1st commit**: adds new search modes for items and folders allowing them to be looked up directly by name and parent ID, which makes it much easier to externally query for the existence of a resource at a given path when the parent ID is known.

* **2nd commit**: some refactoring to reduce code duplication in our swagger descriptions.

* **3rd commit**: move some private methods in girder client into the public API and tweak them a little for consistency. These use the new lookup mode. Also added tests that exercise all this new functionality.

### CODE REVIEW

File: girder/api/v1/folder.py [hunk 1]
```diff
@@ -77,10 +77,12 @@ def find(self, params):
                 exc=True)
```
Comment (brianhelba) [inline girder/api/v1/folder.py]: You should make a note in the docstring on line 54 about name-match search.

File: girder/api/v1/item.py [hunk 2]
```diff
@@ -68,10 +68,13 @@ def find(self, params):
             folder = self.model('folder').load(id=params['folderId'], user=user,
```
Comment (brianhelba) [inline girder/api/v1/item.py]: You should make a note in the docstring on line 51 about name-match search.

**Comment by brianhelba** [comment]
LGTM other than maybe adding some extra internal documentation.

+1 on the change making those Girder client methods public.

**Comment by zachmullen** [comment]
> +1 on the change making those Girder client methods public.

This was actually the main purpose of this branch; the TCGA ingest script currently depends on this functionality. I'll merge once travis is happy.

**Pull request merged**